### ISSUE TITLE
Image quality is low

### ISSUE THREAD

**Issue by VariabileAleatoria** [kevinzg/facebook-scraper#6]
Hi, the link of 'image' returned by get_posts is different from the link you get with 'show image' on browser and it gives back an image on a lower resolution.
Would it be possible to change it in order to get the highest resolution?

**Comment by kevinzg** [comment]
@lazanet changed the image extraction code on #5 to get high quality images. I think there are some cases where it still gets the low res version, but you can try it out.

**Comment by Sarun1001** [comment]
Hi, thanks for this project,
but i also have this request for high-quality images,  most of the current images returned are low res (s320x320)
eg : https://scontent.fblr1-4.fna.fbcdn.net/v/t1.0-0/cp0/e15/q65/s320x320/70134960_2395812457345024_3129842513699930112_n.jpg?_nc_cat=105&_nc_oc=AQlMQ_yg9AeDhAyhvkDVWBdhB4IkkxuSJ-89xNNn3AS8SIYn3vghX7SGRafCLcNsZfg&_nc_ht=scontent.fblr1-4.fna&oh=52f8a7c48a8064bfdd1f1b6b7d05d5c0&oe=5DF96A81

is it possible to retrieve original images

**Comment by jwesheath** [comment]
I made an attempt at fixing this and will submit a pull request.

I believe the issue is with the _photo_link regex.  When looking at these links in Chrome dev tools, they have the form '<a href="(url we want)" class="stuff">', but in the html returned by the request in the code they have the form '<a class="stuff" href="(url we want)">'.  The _photo_link regex does not expect the class attribute to be between the a tag and the href, so it misses the link and falls back to the low quality image.  Removing the a tag from the regex and just looking for the href seems to fix the problem.

**Comment by kevinzg** [comment]
Thank you @jwesheath. @VariabileAleatoria or @Sarun1001 can you confirm the images are now in high resolution?

**Comment by Sarun1001** [comment]
@kevinzg @jwesheath No image quality is still low

**Comment by jwesheath** [comment]
@Sarun1001 are you sure you're looking at the mobile version of Facebook?  Could you please provide an example of a page where you're not getting high res images?  I just tried a couple of pages with the latest version of facebook-scraper and I seem to be getting high res images.

**Comment by Sarun1001** [comment]
@jwesheath Can you try this page [](https://www.facebook.com/Troll.Malayalam/)

**Comment by jwesheath** [comment]
When I run get_posts() on that page, the image link that is returned for the post you mention is for an image that is 653x736 pixels, which seems to be the highest resolution available for that image.  Are you sure you're using the latest version?

**Comment by Sarun1001** [comment]
@jwesheath hi, can you please post the response here, I am still getting low-quality images,
 currently running version 0.1.5
Thank you

**Comment by jwesheath** [comment]
@Sarun1001 The example you posted before is far enough down the page that it's difficult to locate, but I have confirmed in multiple environments and on multiple Facebook pages that version 0.1.5 is getting the high res images.  I'm not saying that you're incorrect - it's entirely possible that my solution is not working properly under certain circumstances, but I'm unable to reproduce the issue so I'm not sure how to proceed.  It would be helpful if we had some more information or some more eyes on this to figure out what the issue is, but in the meantime I'll have a look and see if I can figure out what might cause an issue in environments that are different from the ones I'm working in.

**Comment by nubpro** [comment]
@Sarun1001 @jwesheath 
Upon painstakingly debugging python code for the first time, I figured that you in order to retrieve full sized image, you need to provide credentials in _get_posts_.

I'd observed that it is possible to retrieve full sized image without prior logging in but that would require some changes to the source.

**Comment by jwesheath** [comment]
@nubpro I'm glad that you've found a workaround that resolves the issue for you, but it's not clear to me how providing credentials would affect this.  I actually added the credentials feature so that the scraper would be able to view private Facebook pages, and I'm able to get full size images from public pages without providing credentials.  Does this mean that you were unable to get full size images when you didn't provide credentials?

**Comment by nubpro** [comment]
@jwesheath 
In order to simulate the environment and to visualize the UI that we cannot see, lets use Postman (I have no idea what better way to show this)

URL: https://m.facebook.com/363117593762031/photos/pcb.3391238207616606/3391439877596439/

**Environment 1**
User-agent is left empty. 
![image](https://user-images.githubusercontent.com/762914/72760884-2de23280-3c15-11ea-9b5b-a0b96050536e.png)
Result:
We can see that the 'Lihat Saiz Penuh' (View Full Size) button and the post information is rendered at the browser.

**Environment 2**
User-agent is set.
![image](https://user-images.githubusercontent.com/762914/72761104-e5774480-3c15-11ea-9c4a-1625d9b87068.png)
Result:
The post description and details aren't available in the DOM. Only the picture can be seen.
Thus, this is what causes the code to break.

**Solution**
Setting cookies to `locale=en_US; noscript=1` would rectify the above problems, refer below.
![image](https://user-images.githubusercontent.com/762914/72761220-4b63cc00-3c16-11ea-945e-ba901d42ba77.png)
Setting _noscript_ to _1_ will tell Facebook servers to render the DOM without any Javascript elements and scripts. I believe this is what the author is trying to achieve originally but he may have overlooked it.

Setting _locale_ to _en_US_, would force Facebook to be rendered in plain English instead of locale language despite _Accept-Language_ is already set to _en-US_. Apparently, Facebook would disregard _Accept-Language_.

--
I propose that the following is modified:
_From_
`_headers = {'User-Agent': _user_agent, 'Accept-Language': 'en-US,en;q=0.5'}`

_To_
`_headers = {'User-Agent': _user_agent, 'Accept-Language': 'en-US,en;q=0.5', 'cookie': 'locale=en_US; noscript=1'}`

The following changes will break the current implementation due to DOM changes, hence we may need to perform minimal maintenace to the regex syntax to get it up and running again.

**Comment by kevinzg** [comment]
@nubpro nice findings, could you submit a pull request?

**Comment by jwesheath** [comment]
@nubpro Just to confirm - the current version of facebook-scraper does not provide full size images for you?

I'm able to reproduce your example in Postman, and I agree that it looks like it could be the source of the problem.  What confuses me is that the current version of facebook-scraper gets the full size image for me on multiple machines/environments.  I understand that "works on my machine" is not a good answer, but I haven't been able to reproduce the problem when using the code as it is currently.  Perhaps User-Agent is getting changed in different environments running the same code?  Maybe the requests_html package is doing something differently in different environments? In any case, I certainly have no issue with you modifying the tiny bit of code I've contributed if it helps others.

**Comment by nubpro** [comment]
> @nubpro Just to confirm - the current version of facebook-scraper does not provide full size images for you?
> 
> I'm able to reproduce your example in Postman, and I agree that it looks like it could be the source of the problem. What confuses me is that the current version of facebook-scraper gets the full size image for me on multiple machines/environments. I understand that "works on my machine" is not a good answer, but I haven't been able to reproduce the problem when using the code as it is currently. Perhaps User-Agent is getting changed in different environments running the same code? Maybe the requests_html package is doing something differently in different environments? In any case, I certainly have no issue with you modifying the tiny bit of code I've contributed if it helps others.

Yes, I'm using 0.1.7.

Here what I think we could try to understand your environment:
1. Provide the facebook page.
2. Provide the header and user agents.
3. Are you authenticated with Facebook? (Also, check your session cookies)
4. Provide the stored data in the html variable (line 171, under _extract_photo_link)

Honestly, I think there is a lot of possibilities, it's going to take lots of effort to find out.

--
I'm baking the pull request as we speak. Ah its gonna hooot.

**Comment by nubpro** [comment]
Hey guys, I would like to make amendments to what I've said in the initial post.

In Test 2, I mentioned that the post information and the 'View full size' button isn't available in the DOM, I was completely wrong. Facebook has hid these part of the code by commenting them out, therefore it wasn't rendering in the browser. 
![image](https://user-images.githubusercontent.com/762914/72931243-79711980-3d98-11ea-914d-6e96fdcdda9a.png)

**Comment by nubpro** [comment]
@VariabileAleatoria @Sarun1001 
Give the new version a shot, 0.1.8 and report back the result.

**Comment by VariabileAleatoria** [comment]
Hi, I know this is an hard topic but I want to point out this behaviour.  
I'm now getting low quality images since I started feeding cookies to the function, without cookies I get full resolution images instead.

**Comment by neon-ninja** [comment]
@VariabileAleatoria which post are you having this problem with?

**Comment by VariabileAleatoria** [comment]
> @VariabileAleatoria which post are you having this problem with?

here an example of getting the same post without or with cookies:
```
>>> list(get_posts('testtelegrambotyolo', pages=2))[0]['image']
sys:1: UserWarning: A low page limit (<=2) might return no results, try increasing the limit
'https://scontent-mxp1-1.xx.fbcdn.net/v/t1.6435-9/180736256_516644253052081_8323589748985846252_n.png?_nc_cat=110&ccb=1-3&_nc_sid=8024bb&_nc_ohc=aEjmub0eUJYAX9KEp1G&_nc_ht=scontent-mxp1-1.xx&oh=669e6679d115db23800e0e93ab3ede35&oe=60AFC653'
>>> list(get_posts('testtelegrambotyolo', pages=2, cookies='cookies.txt'))[0]['image']
'https://scontent-mxp1-1.xx.fbcdn.net/v/t1.6435-0/p320x320/180736256_516644253052081_8323589748985846252_n.png?_nc_cat=110&ccb=1-3&_nc_sid=8024bb&efg=eyJpIjoidCJ9&_nc_ohc=aEjmub0eUJYAX9KEp1G&_nc_ht=scontent-mxp1-1.xx&tp=30&oh=5228c44fef3dd7dd2fb2af0b421a94f4&oe=60B2BB40'
```

**Comment by neon-ninja** [comment]
When I tried that:

```python
>>> list(get_posts('testtelegrambotyolo', pages=2))[0]['image']
sys:1: UserWarning: A low page limit (<=2) might return no results, try increasing the limit
'https://scontent.fakl1-3.fna.fbcdn.net/v/t1.6435-9/180736256_516644253052081_8323589748985846252_n.png?_nc_cat=110&ccb=1-3&_nc_sid=8024bb&_nc_ohc=aEjmub0eUJYAX-fGUJJ&_nc_ht=scontent.fakl1-3.fna&oh=5d3a7a838304b71e9763afbcbd0ae67b&oe=60AFC653'
>>> list(get_posts('testtelegrambotyolo', pages=2, cookies='cookies.txt'))[0]['image']
'https://scontent.fakl1-3.fna.fbcdn.net/v/t1.6435-9/180736256_516644253052081_8323589748985846252_n.png?_nc_cat=110&ccb=1-3&_nc_sid=8024bb&efg=eyJpIjoidCJ9&_nc_ohc=aEjmub0eUJYAX-fGUJJ&_nc_ht=scontent.fakl1-3.fna&oh=5d3a7a838304b71e9763afbcbd0ae67b&oe=60AFC653'
````

It's difficult for me to debug this when I get different results to you.

**Comment by VariabileAleatoria** [comment]
> When I tried that:
> 
> ```python
> >>> list(get_posts('testtelegrambotyolo', pages=2))[0]['image']
> sys:1: UserWarning: A low page limit (<=2) might return no results, try increasing the limit
> 'https://scontent.fakl1-3.fna.fbcdn.net/v/t1.6435-9/180736256_516644253052081_8323589748985846252_n.png?_nc_cat=110&ccb=1-3&_nc_sid=8024bb&_nc_ohc=aEjmub0eUJYAX-fGUJJ&_nc_ht=scontent.fakl1-3.fna&oh=5d3a7a838304b71e9763afbcbd0ae67b&oe=60AFC653'
> >>> list(get_posts('testtelegrambotyolo', pages=2, cookies='cookies.txt'))[0]['image']
> 'https://scontent.fakl1-3.fna.fbcdn.net/v/t1.6435-9/180736256_516644253052081_8323589748985846252_n.png?_nc_cat=110&ccb=1-3&_nc_sid=8024bb&efg=eyJpIjoidCJ9&_nc_ohc=aEjmub0eUJYAX-fGUJJ&_nc_ht=scontent.fakl1-3.fna&oh=5d3a7a838304b71e9763afbcbd0ae67b&oe=60AFC653'
> ```
> 
> It's difficult for me to debug this when I get different results to you. What version are you using @VariabileAleatoria?

Version: 0.2.29

**Comment by VariabileAleatoria** [comment]
with last version from git I get full resolution without cookies and None with cookies (and the low resolution one as image_lowquality)

**Comment by neon-ninja** [comment]
What language do you have set on your account?

**Comment by VariabileAleatoria** [comment]
> What language do you have set on your account?

Italian, I just tried switching my account to english and it looks like it's working (even though the locale cookie is still it_IT).  
That's curious, I'll keep testing it and let you know the behaviour!

**Comment by neon-ninja** [comment]
> > What language do you have set on your account?
> 
> Italian, I just tried switching my account to english and it looks like it's working (even though the locale cookie is still it_IT).
> That's curious, I'll keep testing it and let you know the behaviour!

Yes, the scraper looks for a link with the text "View full size", which "Visualizza a grandezza naturale" won't matcj

**Comment by neon-ninja** [comment]
@VariabileAleatoria following https://github.com/kevinzg/facebook-scraper/issues/281 and https://github.com/kevinzg/facebook-scraper/commit/77c7fb6b2c1103affefbf9b690a13f48a53bd031, view full size link extraction should now work in other languages
